torch/csrc/autograd/engine.cpp

```diff
@@ -742,44 +752,134 @@ auto Engine::execute(const edge_list& roots,
+    // would notify the owner thread that the task has been completed.
+    mark_graph_task_completed(graph_task);
+    LOG(INFO) <<"finished graph_task: " << graph_task;
+  } else {
+    // schedule a continuation
+    at::launch([this, graph_task]() {
+        execute_until_ready_queue_empty(graph_task);
+    });
+    LOG(INFO) << "scheduled a continuation on graph_task: "  << graph_task
+      << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+
 }
 
 std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
     const std::shared_ptr<GraphTask>& graph_task,
-    std::shared_ptr<Node> graph_root) {
-  initialize_threads_pool();
+    std::shared_ptr<Node> graph_root,
+    bool async_mode) {
+  initialize_device_threads_pool();
   // Lock mutex for GraphTask.
   std::unique_lock<std::mutex> lock(graph_task->mutex_);
 
-  ready_queue(at::kCPU).push(
+  ready_queue(graph_task, at::kCPU).push(
       NodeTask(graph_task, std::move(graph_root), InputBuffer(0)));
 
-  // Not a worker
+  // worker_device == NO_DEVICE it's a CPU thread and it's trying to drive the
+  // autograd engine with corresponding GraphTask, and its NOT a re-entrant call
   if (worker_device == NO_DEVICE) {
     // graph_task_exec_post_processing is done when the Future is marked as
     // completed in mark_graph_task_completed.
-    return graph_task->future_result_;
+
+    // We set the worker_device to CPU_DEVICE only if worker_device was previously
+    // NO_DEVICE. Setting it to CPU afterwards allow us to detect whether this is
+    // a re-entrant call or not.
+    //
+    // If worker_device = NO_DEVICE: this is NOT a re-entrant call
+    // If worker_device is other devices (i.e. CPU, CUDA): this is a re-entrant
+    //    backward call from that device.
+    set_device(CPU_DEVICE);
+
+    // set the graph_task owner to the current device
+    graph_task->owner_ = worker_device;
+
+    // The owning thread start to drive the engine execution with the GraphTask
+    // that has already been pushed to the current CPU thread's ready_queue
+    lock.unlock();
+    if (async_mode) {
+      execute_until_ready_queue_empty(graph_task);
+    } else {
+      thread_main(nullptr, false);
+      TORCH_INTERNAL_ASSERT(graph_task->future_result_->completed());
+    }
+    // reset the worker_device after the completion of the graph_task, this is so
+    // that the initial state of the engine remains the same across every backward()
+    // or grad() call, we don't need to reset local_ready_queue as we could possibly
+    // reuse it for new backward calls.
+    worker_device = NO_DEVICE;
   } else {
+    // this must be a re-entrant call from worker_device
     graph_task->owner_ = worker_device;
     if (current_depth >= max_recursion_depth_) {
       // See Note [Reentrant backwards]
       // If reached the max depth, switch to a different thread
+      lock.unlock();
```

Comment:
Why is this unlock needed now?

Reply:
I guess `add_thread_pool_task` doesn't need the lock for graph_task, I'm assuming this is a minor lock optimization?